### PR TITLE
methodimpl(.override) should be accompanied by [PreserveBaseOverridesAttribute]

### LINKED ISSUE #45372: methodimpl(.override) should be accompanied by [PreserveBaseOverridesAttribute]
Whenever the compiler produces a `methodimpl` (`.override`) entry in metadata for an override of a class method, the method for which it is produced should have the attribute `System.Runtime.CompilerServices.PreserveBaseOverridesAttribute` (formerly known as `RequireMethodImplToRemainInEffectAttribute`) applied to it to ensure correct behavior in binary compatibility scenarios. The possible exception to this is the `methodimpl` produced for a destructor declaration to indicate that it overrides the corresponding method in `object`; for that the attribute is probably not necessary (but also not harmful).

### PR DESCRIPTION
**Pull request by gafter**
Fixes #45372
Relates to #44208

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs [hunk 1]
```diff
@@ -874,10 +874,9 @@ internal override void AddSynthesizedAttributes(PEModuleBuilder moduleBuilder, r
                 AddSynthesizedAttribute(ref attributes, moduleBuilder.SynthesizeNullableContextAttribute(this, nullableContextValue));
             }
 
-            bool isCovariantOverride = this.IsOverride && !this.OverriddenMethod.ReturnType.Equals(this.ReturnType, TypeCompareKind.AllIgnoreOptions);
-            if (isCovariantOverride)
+            if (this.RequiresExplicitOverride(out _))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs:877]: >this.RequiresExplicitOverride(out _) [](start = 16, length = 36)

I do not believe this is we would want to apply the attribute across the board (especially for cases for which we emitted MethodImpl in previous versions of the compiler), but rather should target true covariant return cases. I also think that for those, the attribute application is not optional, but required.

**Comment by AlekseyTs** [comment]
I think we should add more targeted tests (less generalized) for different scenarios, especially for those legacy scenarios that already used MethodImple prior to any changes made around covariant returns. The adjustment to the test helper to add additional verification is, of course, good. However, it is almost impossible to judge what scenarios are actually covered.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs:877]: This PR implements [https://github.com/dotnet/roslyn/issues/45372](https://github.com/dotnet/roslyn/issues/45372). This is intentional. The behavior of the methodimpl in the absence of this attribute does not match the language semantics, particularly in binary compat scenarios. On advice of David Wrighton, when it is available this attribute is used on all methodimpl for a class member override as described in [https://github.com/dotnet/roslyn/issues/45372](https://github.com/dotnet/roslyn/issues/45372).

---
In reply to: [452991537](https://github.com/dotnet/roslyn/pull/45870#discussion_r452991537) [](ancestors = 452991537)

**Comment by gafter** [comment]
I've added tests for both scenarios in which the prior compiler produced methodimpl (property skipping a matching method in an intervening base type and vice versa). All of the rest of the scenarios are new ones and are exercised by existing tests.

---
In reply to: [656810422](https://github.com/dotnet/roslyn/pull/45870#issuecomment-656810422) [](ancestors = 656810422)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs:877]: What to do when the attribute is missing is an open issue recorded in [https://github.com/dotnet/roslyn/issues/44208](https://github.com/dotnet/roslyn/issues/44208)

---
In reply to: [453862057](https://github.com/dotnet/roslyn/pull/45870#discussion_r453862057) [](ancestors = 453862057,452991537)

**Comment by gafter** [comment]
@AlekseyTs I think I've responded to all of your comments. Do you have any others?
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Pull request merged**